pkg/ec2provider/ec2provider.go

```diff
@@ -0,0 +1,274 @@
+	default_wait_interval = 50
+	// Making sure the single instance calls waits max till 5 seconds 100* (50 * time.Millisecond)
+	total_iteration_for_wait_interval = 100
+	// Maximum number of instances with which ec2:DescribeInstances call will be made
+	max_instances_batch_size = 100
+	// Maximum time to wait for a new batch call this also depends on if the instance size has
+	// already become 100 then it will not respect this limit
+	max_wait_interval_for_batch = 200
+)
+
+type EC2Provider interface {
+	// Get a node name from instance ID
+	GetPrivateDNSName(string) (string, error)
+	StartEc2DescribeBatchProcessing()
+}
+
+type ec2PrivateDNSCache struct {
+	dnsCache     map[string]string
+	dnsCacheLock sync.RWMutex
+}
+
+type ec2RequestQueue struct {
+	requestQueueMap  map[string]bool
+	requestQueueLock sync.RWMutex
+}
+
+type ec2ProviderImpl struct {
+	ec2                ec2iface.EC2API
+	privateDNSCache    ec2PrivateDNSCache
+	requestQueue       ec2RequestQueue
+	instanceIdsChannel chan string
+}
+
+func NewEC2Provider(roleARN string, qps int, burst int) EC2Provider {
+	dnsCache := ec2PrivateDNSCache{
+		dnsCache:     make(map[string]string),
+		dnsCacheLock: sync.RWMutex{},
+	}
+	ec2RequestQueue := ec2RequestQueue{
+		requestQueueMap:  make(map[string]bool),
+		requestQueueLock: sync.RWMutex{},
+	}
+	return &ec2ProviderImpl{
+		ec2:                ec2.New(newSession(roleARN, qps, burst)),
+		privateDNSCache:    dnsCache,
+		requestQueue:       ec2RequestQueue,
+		instanceIdsChannel: make(chan string, max_channel_size),
+	}
+}
+
+func newSession(roleARN string, qps int, burst int) *session.Session {
+	// Initial credentials loaded from SDK's default credential chain, such as
+	// the environment, shared credentials (~/.aws/credentials), or EC2 Instance
+	// Role.
+
+	sess := session.Must(session.NewSession())
+	if aws.StringValue(sess.Config.Region) == "" {
+		ec2metadata := ec2metadata.New(sess)
+		regionFound, err := ec2metadata.Region()
+		if err != nil {
+			logrus.WithError(err).Fatal("Region not found in shared credentials, environment variable, or instance metadata.")
+		}
+		sess.Config.Region = aws.String(regionFound)
+	}
+
+	if roleARN != "" {
+		logrus.WithFields(logrus.Fields{
+			"roleARN": roleARN,
+		}).Infof("Using assumed role for EC2 API")
+
+		rateLimitedClient, err := httputil.NewRateLimitedClient(qps, burst)
+
+		if err != nil {
+			logrus.Errorf("Getting error = %s while creating rate limited client ", err)
+		}
+
+		ap := &stscreds.AssumeRoleProvider{
+			Client:   sts.New(sess, aws.NewConfig().WithHTTPClient(rateLimitedClient).WithSTSRegionalEndpoint(endpoints.RegionalSTSEndpoint)),
+			RoleARN:  roleARN,
+			Duration: time.Duration(60) * time.Minute,
+		}
+
+		sess.Config.Credentials = credentials.NewCredentials(ap)
+	}
+	return sess
+}
+
+func (p *ec2ProviderImpl) setPrivateDNSNameCache(id string, privateDNSName string) {
+	p.privateDNSCache.dnsCacheLock.Lock()
+	defer p.privateDNSCache.dnsCacheLock.Unlock()
```

Comment:
This is good! Change to use this "lock / defer unlock"-pattern in all the functions attached to `ec2ProviderImpl`.

Reply:
Yeah , I like your idea as I was trying to be performance agnostic about these code in terms of nanoseconds while loosing the readability.
I will make all the unlock with defer.